ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -450,11 +392,145 @@ private AutomatonPerimeterResultImpl simulateAutomatonState(State automatonState
+        if (!appliedNetworkActions.isEmpty()) {
+            TECHNICAL_LOGS.info("Running sensi post application of auto network actions for automaton state {}.", automatonState.getId());
+            automatonRangeActionOptimizationSensitivityAnalysisOutput = preAutoPerimeterSensitivityAnalysis.runBasedOnInitialResults(network, initialFlowResult, operatorsNotSharingCras, null);
+            RaoLogger.logMostLimitingElementsResults(TECHNICAL_LOGS, automatonRangeActionOptimizationSensitivityAnalysisOutput, raoParameters.getObjectiveFunction(), NUMBER_LOGGED_ELEMENTS_DURING_RAO);
+        }
+
+        return Pair.of(automatonRangeActionOptimizationSensitivityAnalysisOutput, appliedNetworkActions);
+    }
+
+    /**
+     * This function gathers the flow cnecs to be considered while shifting range actions,
+     * depending on the range action availableRa's UsageMethod.
+     */
+    static Set<FlowCnec> gatherFlowCnecs(RangeAction<?> availableRa,
+                                           State automatonState,
+                                           Crac crac,
+                                           Network network) {
+        // UsageMethod is either FORCED or TO_BE_EVALUATED
+        // -- FORCED :
+        Set<FlowCnec> flowCnecs = crac.getFlowCnecs().stream()
+                .filter(flowCnec -> flowCnec.getState().equals(automatonState))
+                .collect(Collectors.toSet());
+        // -- TO_BE_EVALUATED :
+        if (availableRa.getUsageMethod(automatonState).equals(UsageMethod.TO_BE_EVALUATED)) {
+            // Get flowcnecs constrained by OnFlowConstraint
+            flowCnecs = availableRa.getUsageRules().stream()
+                    .filter(OnFlowConstraint.class::isInstance)
+                    .map(OnFlowConstraint.class::cast)
+                    .map(OnFlowConstraint::getFlowCnec)
+                    .filter(flowCnec -> flowCnec.getState().equals(automatonState))
+                    .collect(Collectors.toSet());
+            // Get all cnecs in country if availableRa is available on a OnFlowConstraintInCountry usage rule
+            Set<Country> countries = availableRa.getUsageRules().stream()
+                    .filter(OnFlowConstraintInCountry.class::isInstance)
+                    .map(OnFlowConstraintInCountry.class::cast)
+                    .map(OnFlowConstraintInCountry::getCountry)
+                    .collect(Collectors.toSet());
+            flowCnecs.addAll(crac.getFlowCnecs().stream()
+                    .filter(flowCnec -> flowCnec.getState().equals(automatonState))
+                    .filter(flowCnec -> {
+                        for (Country country : countries) {
+                            if (RaoUtil.isCnecInCountry(flowCnec, country, network)) {
+                                return true;
+                            }
+                        }
+                        return false;
+                    })
```

Comment:
you can replace this with filter(flowCnec -> countries.stream().anyMatch(RaoUtil.isCnecInCountry(flowCnec, country, network)))